node/lib/util/stitch_util.js

```diff
@@ -52,41 +53,124 @@ const FILEMODE            = NodeGit.TreeEntry.FILEMODE;
 
 const maxParallel = 1000;
 
+
+// TODO: the `writeNotes` and `readNotes` methods should be moved to a utility
+// for notes, if they're needed elsewhwere.
+
 /**
- * The name of the note used to record conversion information.
+ * Write the specified `contents` to the note having the specified `refName` in
+ * the specified `repo`.
  *
- * @property {String}
+ * Writing notes oneo-at-a-time is slow.  This method let's you write them in
+ * bulk, far more efficiently.
+ *
+ * @param {NodeGit.Repository} repo
+ * @param {String}             refName
+ * @param {Object}             contents    SHA to data
  */
-exports.convertedNoteRef = "refs/notes/stitched/converted";
+exports.writeNotes = co.wrap(function *(repo, refName, contents) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isString(refName);
+    assert.isObject(contents);
+
+    if (0 === Object.keys(contents).length) {
+        // Nothing to do if no contents; no point in making an empty commit or
+        // in making clients check themselves.
+        return;                                                       // RETURN
+    }
+
+    // We're going to directly write the tree/commit for a new note containing
+    // `contents`.
+
+    let currentCommit = null;
+    let currentTree = null;
+    const parents = [];
+    const ref = yield GitUtil.getReference(repo, refName);
+    if (null !== ref) {
+        currentCommit = yield repo.getCommit(ref.target());
+        parents.push(currentCommit);
+        currentTree = yield currentCommit.getTree();
+    }
+    const odb = yield repo.odb();
+    const changes = {};
+    const ODB_BLOB = 3;
+    const BLOB = NodeGit.TreeEntry.FILEMODE.BLOB;
+    const writeBlob = co.wrap(function *(sha) {
+        const content = contents[sha];
+        const blobId = yield odb.write(content, content.length, ODB_BLOB);
+        changes[sha] = new TreeUtil.Change(blobId, BLOB);
```

Comment:
That is not how notes trees are supposed to work -- there's a whole sharding scheme.  Isn't there something in NodeGit that does this?

Reply:
Git seems to be happy with it in that I can read notes written this way with `git notes show`. However, you're right, I should change to use the sharding scheme.

libgit2 does not provide a way to read or write more than one note at a time; these bulk operations provided a significant performance improvement.